core/commands/pin.go

```diff
@@ -282,3 +272,76 @@ type RefKeyObject struct {
 type RefKeyList struct {
 	Keys map[string]RefKeyObject
 }
+
+func pinLsKeys(args []string, typeStr string, ctx context.Context, n *core.IpfsNode) (map[string]RefKeyObject, error) {
+
+	keys := make(map[string]RefKeyObject)
+
+	for _, p := range args {
+		dagNode, err := core.Resolve(ctx, n, path.Path(p))
+		if err != nil {
+			return nil, err
+		}
+
+		k, err := dagNode.Key()
+		if err != nil {
+			return nil, err
+		}
+
+		pinType, pinned, err := n.Pinning.IsPinnedWithType(k, typeStr)
+		if err != nil {
+			return nil, err
+		}
+
+		if !pinned {
+			return nil, fmt.Errorf("Path '%s' is not pinned", p)
+		}
+
+		switch pinType {
+		case "direct", "indirect", "recursive", "internal":
+		default:
+			pinType = "indirect through " + pinType
+		}
+		keys[k.B58String()] = RefKeyObject{
+			Type: pinType,
+		}
+	}
+
+	return keys, nil
+}
+
+func pinLsAll(typeStr string, ctx context.Context, n *core.IpfsNode) (map[string]RefKeyObject, error) {
+
+	keys := make(map[string]RefKeyObject)
+
+	AddToResultKeys := func(keyList []key.Key, typeStr string) {
+		for _, k := range keyList {
+			keys[k.B58String()] = RefKeyObject{
+				Type: typeStr,
+			}
+		}
+	}
+
+	if typeStr == "direct" || typeStr == "all" {
+		AddToResultKeys(n.Pinning.DirectKeys(), "direct")
+	}
+	if typeStr == "indirect" || typeStr == "all" {
+		ks := key.NewKeySet()
+		for _, k := range n.Pinning.RecursiveKeys() {
+			nd, err := n.DAG.Get(ctx, k)
+			if err != nil {
+				return nil, err
+			}
+			err = dag.EnumerateChildren(n.Context(), n.DAG, nd, ks)
```

Comment:
Listing indirect keys recursively appears only in here, while I think it should the case in `pinLsKeys`'s indirect as well.